packages/flutter/lib/src/material/scrollbar.dart

```diff
@@ -2,10 +2,186 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
 import 'package:flutter/widgets.dart';
 
 import 'theme.dart';
 
+class Scrollbar2 extends StatefulWidget {
+  Scrollbar2({
+    Key key,
+    this.child,
+  }) : super(key: key);
+
+
+  /// The subtree to place inside the [Scrollbar2]. This should include
+  /// a source of [ScrollNotification2] notifications, typically a [Scrollable2]
+  /// widget.
+  final Widget child;
+
+  @override
+  _Scrollbar2State createState() => new _Scrollbar2State();
+}
+
+class _Scrollbar2State extends State<Scrollbar2> with TickerProviderStateMixin {
+  _ScrollbarController _controller;
+
+  @override
+  void dependenciesChanged() {
+    super.dependenciesChanged();
+    _controller ??= new _ScrollbarController(this);
+    _controller.color = Theme.of(context).highlightColor;
+  }
+
+  bool _handleScrollNotification(ScrollNotification2 notification) {
+    if (notification is ScrollUpdateNotification ||
+        notification is OverscrollNotification)
+      _controller.update(notification.metrics, notification.axisDirection);
+    return false;
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new NotificationListener<ScrollNotification2>(
+      onNotification: _handleScrollNotification,
+      child: new RepaintBoundary(
+        child: new CustomPaint(
+          foregroundPainter: new _ScrollbarPainter(_controller),
+          child: new RepaintBoundary(
+            child: config.child,
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class _ScrollbarController extends ChangeNotifier {
+  _ScrollbarController(TickerProvider vsync) {
+    assert(vsync != null);
+    _fadeController = new AnimationController(duration: _kThumbFadeDuration, vsync: vsync);
+    _opacity = new CurvedAnimation(parent: _fadeController, curve: Curves.fastOutSlowIn)
+      ..addListener(notifyListeners);
+  }
+
+  // animation of the main axis direction
+  AnimationController _fadeController;
+  Animation<double> _opacity;
+
+  // fade-out timer
+  Timer _fadeOut;
+
+  Color get color => _color;
+  Color _color;
+  set color(Color value) {
+    assert(value != null);
+    if (color == value)
+      return;
+    _color = value;
+    notifyListeners();
+  }
+
+  @override
+  void dispose() {
+    _fadeOut?.cancel();
+    _fadeController.dispose();
+    super.dispose();
+  }
+
+  ScrollableMetrics _lastMetrics;
+  AxisDirection _lastAxisDirection;
+
+  static const double _kMinThumbExtent = 18.0;
+  static const double _kThumbGirth = 6.0;
+  static const Duration _kThumbFadeDuration = const Duration(milliseconds: 300);
+  static const Duration _kFadeOutTimeout = const Duration(milliseconds: 600);
```

Comment:
For private names like these, I've just been putting them as file-level private constants.